packages/frontend/app/routes/tenants.$tenantId.tsx

```diff
@@ -0,0 +1,321 @@
+        {/* Sensitive Info */}
+        <div className='grid grid-cols-1 py-3 gap-6 md:grid-cols-3 border-b border-pearl'>
+          <div className='col-span-1 pt-3'>
+            <h3 className='text-lg font-medium'>Sensitive Information</h3>
+            <ErrorPanel errors={response?.errors.message} />
+          </div>
+          <div className='md:col-span-2 bg-white rounded-md shadow-md'>
+            <Form method='post' replace preventScrollReset>
+              <fieldset disabled={isSubmitting}>
+                <div className='w-full p-4 space-y-3'>
+                  <Input type='hidden' name='id' value={tenant.id} />
+                  <PasswordInput
+                    name='apiSecret'
+                    label='API Secret'
+                    value={tenant.apiSecret}
+                    required
+                    disabled
+                  />
+                </div>
+              </fieldset>
+            </Form>
+          </div>
+        </div>
+        {/* Sensitive - END */}
+        {/* DELETE TENANT - Danger zone */}
+        {me.isOperator && me.id !== tenant.id && (
+          <DangerZone title='Delete Tenant'>
+            <Form method='post' onSubmit={submitHandler}>
+              <Input type='hidden' name='id' value={tenant.id} />
+              <Input type='hidden' name='intent' value='delete' />
+              <Button type='submit' intent='danger' aria-label='delete tenant'>
+                Delete tenant
+              </Button>
+            </Form>
+          </DangerZone>
+        )}
+        <ConfirmationDialog
+          ref={dialogRef}
+          onConfirm={onConfirm}
+          title={`Delete Tenant ${tenant.publicName}`}
+          keyword={'delete tenant'}
+          confirmButtonText='Delete this tenant'
+        />
+      </div>
+      <Outlet />
+    </div>
+  )
+}
+
+export async function action({ request }: ActionFunctionArgs) {
+  const errors: {
+    fieldErrors: ZodFieldErrors<typeof createTenantSchema>
+    message: string[]
+  } = {
+    fieldErrors: {},
+    message: []
+  }
+
+  const session = await messageStorage.getSession(request.headers.get('cookie'))
+  const formData = await request.formData()
+  const intent = formData.get('intent')
+  formData.delete('intent')
+
+  switch (intent) {
+    case 'general':
+    case 'ip':
+    case 'sensitive': {
+      const formEntries = Object.fromEntries(formData)
+      const result = updateTenantSchema.safeParse(formEntries)
+      if (!result.success) {
+        errors.fieldErrors = result.error.flatten().fieldErrors
+        return json({ ...errors }, { status: 400 })
+      }
+      if (
+        result.data.email &&
+        result.data.email.trim().length > 0 &&
+        !new RegExp(/^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/).test(
+          result.data.email
+        )
+      ) {
+        errors.fieldErrors.email = ['Email is invalid.']
+        return json({ errors }, { status: 400 })
+      }
```

Comment:
Does `const result = updateTenantSchema.safeParse(formEntries)` take care of checking the email validation already?

Reply:
No, there is an issue with the optional vs regEx, that always forces the regex even when the field is marked as optional. 
As a workaround, I am validating it as part of the `action`.

Reply:
As indicated by @mkurapov  
https://github.com/colinhacks/zod/issues/310

We have a workaround:
```
[`](email: z.string().email().or(z.literal('')),)
```